Add mode options to `chirptime` and `interpolated_tf`

mj-will opened the pull request:
@ConWea @WuShichao @ahnitz we're going to need these changes for the pre-merger paper, so @spxiwh and I agreed that I'd try and get https://github.com/gwastro/BBHX-waveform-model/pull/2 ready while Connor focuses on writing.

This PR updates https://github.com/gwastro/BBHX-waveform-model/pull/2 so that `chirptime` and `interploated_tf` support a `mode` option.

I think I've followed the logic https://arxiv.org/abs/2005.08830, but I may have missed something.

spxiwh commented on BBHX_PhenomD.py:
```diff
@@ -33,6 +39,25 @@ def bbhx_fd(ifos=None, run_phenomd=True, nyquist_freq=0.1,
+    else:
+        err_msg = f"Must provide 'f_lower' or 't_obs_start'."
+
+    wave_gen = get_waveform_genner(mf_min=f_min*MTSUN_SI*(m1+m2), run_phenomd=run_phenomd)
```
I note that in this case mf_min will always have different values, and therefore we should *not* use lru_cache because it won't be able to do anything. Maybe we could lock mf_min to nearest order of magnitude (or half order-of-magnitude), but it might be easy to just ignore this and remove the caching of the waveform_generator

spxiwh commented on BBHX_PhenomD.py:
```diff
@@ -33,6 +39,25 @@ def bbhx_fd(ifos=None, run_phenomd=True, nyquist_freq=0.1,
+        tf_track = interp1d(t_array, freq_array)
+        f_min = tf_track(t_obs_start) # in Hz
+    else:
+        err_msg = f"Must provide 'f_lower' or 't_obs_start'."
```
I think you get this message also if you provide *both*

WuShichao replied:
Yes, I have noticed that, in that case, I think we should force users just set one of them, otherwise we need to choose one of them (and it's possible their `f_min` are different).

spxiwh commented on BBHX_PhenomD.py:
```diff
@@ -1,25 +1,31 @@
+    wave_gen = BBHWaveformFD(amp_phase_kwargs=dict(run_phenomd=run_phenomd, mf_min=mf_min))
     return wave_gen
 
+@functools.lru_cache()
```
Shouldn't be cached assuming m1/m2 are different every time.

WuShichao replied:
@spxiwh According to this example https://docs.python.org/3/library/functools.html it seems it can receive a variable. If not, are there any other methods to accelerate?

spxiwh replied:
@WuShichao This can receive variable inputs, but in the case that the inputs are always different (ie. you will never send the same combination of m1, m2 and f_lower), it will not do anything.

*If* you wanted to have a speed improvement from caching you would need to do something like rounding the masses to some power of ten (or fraction thereof).

I think for the chirptime computation (which is very quick anyway), this is unnecessary complication, so don't use lru_cache here.

The call to initialize BBHWaveformFD was not quick, but is much faster than the waveform generation. We were getting a speedup by using lru_cache there, but now the call is complicated by having a different mf_min every time. My initial thought is that we should also remove the lru_cache there as well, and separately we could consider a more complicated caching routine here, if it is needed.

WuShichao commented on BBHX_PhenomD.py:
```diff
@@ -33,6 +39,25 @@ def bbhx_fd(ifos=None, run_phenomd=True, nyquist_freq=0.1,
+        for freq in freq_array:
+            t_array.append(chirptime(m1=m1, m2=m2, f_lower=freq)/YRSID_SI)
+        tf_track = interp1d(t_array, freq_array)
+        f_min = tf_track(t_obs_start) # in Hz
```
In my first implementation, I used the `phenomhm.tf` from BBHx to calculate the T-F track https://github.com/ConWea/BBHX-waveform-model/pull/2/commits/b0f37f264b6b76c99fa40879350f07a31534b898#diff-e1cfc5e431679fd715906b0736189454cd43ababe331a5c6def3cb416e040b50R42, but Alex said we shouldn't generate entire waveform to get it, so I use the PN equation for chirp time from PyCBC https://github.com/gwastro/pycbc/blob/master/pycbc/waveform/spa_tmplt.py#L32 I think it's faster.

ahnitz commented on BBHX_PhenomD.py:
```diff
@@ -33,6 +39,25 @@ def bbhx_fd(ifos=None, run_phenomd=True, nyquist_freq=0.1,
     psi = params['polarization']
     t_ref = params['tc']
 
+    if 'f_lower' in params and 't_obs_start' not in params:
```
@WuShichao You may want to check how this behaves in practice. I think f_lower might always be in the dict, but if it is 0 or -1 it should probably be treated as if it weren't.

ahnitz commented on BBHX_PhenomD.py:
```diff
@@ -33,6 +39,25 @@ def bbhx_fd(ifos=None, run_phenomd=True, nyquist_freq=0.1,
+        # Using findchirp_chirptime in PyCBC to calculate 
+        # the time-frequency track of dominant mode to get
+        # the corresponding `f_min` for `t_obs_start`.
+        freq_array = np.logspace(0.0001, 1, num=100)
```
What if someone sets f_lower and t_obs_start? Should that not also be OK?

WuShichao replied:
@ahnitz In that case, `f_lower` and `t_obs_start` may corresponding to different `f_min`. Maybe it would be better just let users set one of them.

ahnitz replied:
The one caveat here is the case of higher modes, were the frequency start doesn't quite have a unique (different for each mode) same mapping to time, so being able to set both would be useful.

WuShichao replied:
@ahnitz That makes sense. In that case, we must choose the min `f_min` from `f_lower` and `t_obs_start`.

spxiwh commented:
@WuShichao Would it be possible to provide a profile graph (https://github.com/jrfonseca/gprof2dot) of generating a waveform with this function after this change, and with the lru_cache calls removed? I do want to understand if this has impacted performance.

spxiwh commented on BBHX_PhenomD.py:
```diff
@@ -1,25 +1,51 @@
+import math
+import numpy as np
+from scipy.interpolate import interp1d
+from bbhx.utils.constants import *
```
I think you're only using `YRSID_SI` here. If so, can you just import that, and not import *

WuShichao replied:
Thanks! I change it to `from bbhx.utils.constants import MTSUN_SI, YRSID_SI`.

spxiwh commented on BBHX_PhenomD.py:
```diff
@@ -1,25 +1,51 @@
-    import numpy as np
     from pycbc.types import FrequencySeries, Array
     from pycbc import pnutils
     from bbhx.utils.transform import LISA_to_SSB
```
Move the BBHx import to the top to be consistent *or* move the other bbhx import to here.

WuShichao replied:
Now I move all bbhx imports to the top:
```
from bbhx.utils.constants import MTSUN_SI, YRSID_SI
from bbhx.waveformbuild import BBHWaveformFD
from bbhx.utils.transform import LISA_to_SSB
```

spxiwh commented on BBHX_PhenomD.py:
```diff
@@ -32,6 +58,41 @@ def bbhx_fd(ifos=None, run_phenomd=True, nyquist_freq=0.1,
+
+    if 'f_lower' in params and params['f_lower'] < 0:
+        # the default value of 'f_lower' in PyCBC is -1.
+        params.pop('f_lower')
```
You *must* copy params before doing this. We can't change input structures here in a way that PyCBC may not expect.

spxiwh replied:
Do we actually need to do this? Could we not combine this logic into line 68?

`if ('f_lower' not in params) or (params['f_lower'] < 0):`

WuShichao replied:
Now I use `nparams = params.copy()` and `if ('f_lower' not in nparams) or (nparams['f_lower'] < 0):`.

spxiwh commented on BBHX_PhenomD.py:
```diff
@@ -32,6 +58,41 @@ def bbhx_fd(ifos=None, run_phenomd=True, nyquist_freq=0.1,
+
+    if 'f_lower' not in params:
+        tf_track = interpolated_tf(m1, m2)
+        t_max = chirptime(m1=m1, m2=m2, f_lower=1e-4)
```
It's hardcoded here that f_lower cannot be smaller than 0.0001 if f_lower is not provided in the params. This might be confusing if you ask for a year-long waveform, but you get something much shorter returned because of this condition.

WuShichao replied:
0.1mHz is the low-frequency cutoff for LISA. The `f_min` of 1 year-long SMBHB signal is shown here https://lisa-ldc.lal.in2p3.fr/static/data/img/Sangria_FD.png which is above 0.1mHz, so I think it's safe for SMBHB. I think SOBHB's typical `f_min` is at the high-frequency end of LISA, the plot in slack channel is 5-years long GW150914, and its `f_min` is above 0.01Hz. So for 1 year-long input, `f_min` will be calculated from `t_obs_start`.

spxiwh replied:
Okay, I would rather than 1E-4 not be hardcoded in here, but I do want this merged, so am happy to consider this again later if needed.

spxiwh commented on BBHX_PhenomD.py:
```diff
@@ -32,6 +58,41 @@ def bbhx_fd(ifos=None, run_phenomd=True, nyquist_freq=0.1,
+        else:
+            f_min = tf_track(t_obs_start) # in Hz
+    else:
+        f_min_input = params['f_lower'] # in Hz
```
I think if fmin is provided as a parameter, that's the fmin that must be used! Perhaps raise a warning (or even an error) if the fmin results in a waveform that is longer than t_obs_start, but I don't think fmin should be changed here ... And I don't think the 0.0001 number should appear here either.

WuShichao replied:
I change this to
```
        if f_min < f_min_tobs:
            err_msg = f"Input 'f_lower' is lower than the value calculated from 't_obs_start'."
```

spxiwh reviewed (changes requested):
Thanks @WuShichao I think I'm just left with minor implementation comments now. Once those are dealt with, I think this is good to merge! Apologies for the delay with this!!

spxiwh commented on BBHX_PhenomD.py:
```diff
@@ -1,37 +1,92 @@
-    a2 = params['spin2z']
-    dist = pnutils.megaparsecs_to_meters(params['distance'])
-    phi_ref = params['coa_phase']
+    nparams = params.copy()
```
As you no longer *change* nparams, the copy is no longer needed, and you can just use `params` directly.

spxiwh reviewed (approved):
I made one more minor request, but I think I would then be happy for this to be merged!

spxiwh commented on BBHX_PhenomD.py:
```diff
@@ -48,27 +103,29 @@ def bbhx_fd(ifos=None, run_phenomd=True, nyquist_freq=0.1,
+        if 'f_final' in nparams and nparams['f_final'] != 0:
+            freqs = cached_arange(0, nparams['f_final'], 1/t_obs_start)
+        else:
+            err_msg = f"Please set 'f_final' in **params."
```
Just ran into this: If you get here, you need to `raise` the error message, otherwise this just carries on and fails a bit further down, without saying that we need to set f_final!

WuShichao replied:
@spxiwh Thanks! I have modified my code. Alex is on holiday now, you can merge this if this is good enough.

ConWea committed:
Initial testing to include higher order modes

ConWea committed:
Edit to include option to use gpus

ConWea committed:
Changes to requested mode returns

ConWea committed:
Restructure to check for modes

ConWea committed:
Testing with higher order modes

ConWea committed:
BBHx testing with higher modes edits.

ConWea committed:
More modes stuff

ConWea committed:
More likelihood tests

ConWea committed:
Hard coded changes to separate modes

ConWea committed:
Testing

ConWea committed:
Working higher modes

ConWea committed:
Edits to include higher order modes

ConWea committed:
Length changes

ConWea committed:
Added the appropriate function for approximant in the imr_duration function

ConWea committed:
Moved error message so prints correctly

ConWea committed:
Update error message

ConWea committed:
Injection changes

WuShichao committed:
fix TIME_OFFSET_20_DEGREES str issue

ahnitz committed:
Merge pull request #3 from gwastro/fix_str_to_float

fix TIME_OFFSET_20_DEGREES str issue

ConWea committed:
Added more customization to the BBHx waveform varaibles

ConWea committed:
Raise exception if mode_array is empty for PhenomHM.

ConWea committed:
Set default mode_array to all available modes in PhenomHM if not specified by user

WuShichao committed:
add TDI-2 support

WuShichao committed:
update version number

WuShichao committed:
update

mj-will committed:
add mode options to `chirptime` and `interpolated_tf`

mj-will committed:
add value error for approximant name

WuShichao commented:
@mj-will Thanks for the PR, I will review it very soon.

mj-will commented:
@WuShichao I've spotted a minor bug in this that I'll fix today

mj-will committed:
ensure mode_array is a list and fix mode argument in functions

mj-will commented:
Ian has asked that rather than merge into the existing PR, I make a new PR instead. Once it is open, I will close this.

WuShichao commented on BBHX_Phenom.py:
```diff
@@ -54,29 +68,47 @@ def imr_duration(**params):
     t_array = np.zeros(len(freq_array))
     for i in range(len(freq_array)):
-        t_array[i] = chirptime(m1=m1, m2=m2, f_lower=freq_array[i])
+        t_array[i] = chirptime(m1=m1, m2=m2, f_lower=freq_array[i], m_mode=m_mode)
```
@mj-will Let's do a time-frequency plot check for all the modes. For example, use this function to calculate all the tf tracks for a SMBHB and a SOBHB, and plot them on two plots (one plot for SMBHB or SOBHB), to see if plots make sense.

WuShichao commented on BBHX_Phenom.py:
```diff
@@ -131,35 +164,37 @@ def _bbhx_fd(ifos=None, run_phenomd=True, ref_frame='LISA',
+    # The mode with the largest m will have the highest frequency
+    # since approximately F_lm = m/2 * F_22 and will therefore reach this
+    # frequency at the earliest time
+    max_m_mode = max([mode[1] for mode in mode_array])
```
@mj-will `f_min` should be the lowest frequency at `t_obs_start` among all modes; please check if the value calculated from here matches that time-frequency plot for both SMBHB and SOBHB cases.

mj-will replied:
@WuShichao after some thought, I realised I had been thinking about this the wrong way round; I was thinking about which mode would have the longest chirp time for a given frequency, e.g. which would have the largest `t_max` in the code, rather than which would have the lowest frequency at a given time. I'll update the code to reflect this.

mj-will committed:
add a basic test for get_fd_det_waveform

mj-will committed:
add basic CI for running tests

mj-will committed:
install pytest

mj-will committed:
add missing whitespace

mj-will committed:
rename to tests

WuShichao committed:
use str

WuShichao committed:
let users set TDI version

WuShichao committed:
Update BBHX_PhenomD.py

WuShichao committed:
Update BBHX_PhenomD.py

mj-will committed:
add `imr_duration` for PhenomHM

Addresses https://github.com/gwastro/BBHX-waveform-model/pull/2#discussion_r1510381020

mj-will committed:
handle case where mode array is None

mj-will committed:
remove redundant loop

mj-will committed:
change logic for m mode

mj-will committed:
make number of interpolation points configurable

mj-will committed:
avoid duplicate code

mj-will committed:
use max m mode & only compute track when needed

mj-will committed:
add doc-string to explain new arguments

mj-will committed:
update comment

mj-will committed:
remove option for `log_mf_min=None` in `get_waveform_genner`

mj-will committed:
handle mode_array=None for HM

WuShichao committed:
Merge pull request #8 from gwastro/add-ci-tests

Add ci tests

mj-will committed:
use mamba for ci

mj-will committed:
remove micromamba version

mj-will committed:
rename to tests

mj-will committed:
trigger CI

mj-will committed:
fix typo in CI env

mj-will committed:
specify shell

WuShichao committed:
use BBHx's L_SI

mj-will committed:
pin lapack version

mj-will committed:
add pycbc master

mj-will committed:
remove comment

WuShichao committed:
Merge pull request #10 from gwastro/fix-ci

Fix CI

WuShichao committed:
add TDI versions

WuShichao committed:
update

WuShichao committed:
fix typo

WuShichao committed:
Merge pull request #5 from gwastro/add_tdi_2

add TDI-2 support

mj-will committed:
add tests for HoM

mj-will committed:
Merge remote-tracking branch 'upstream/main' into udpdate-hm-pr

mj-will committed:
add TDI version to sequence test

WuShichao commented:
Do we need to close this one?

mj-will commented:
This PR is no longe needed now that https://github.com/gwastro/BBHX-waveform-model/pull/7 is merged
